lib/explorer/shared.ex

```diff
@@ -37,7 +37,20 @@ defmodule Explorer.Shared do
     if maybe_dtype = normalise_dtype(inner), do: {:list, maybe_dtype}
   end
 
-  def normalise_dtype(dtype) when dtype in @non_list_types, do: dtype
+  def normalise_dtype({:struct, inner_types}) do
+    normalized_dtypes =
+      for {key, dtype} <- inner_types, into: %{} do
+        {key, normalise_dtype(dtype)}
+      end
```

Comment:
You can use `Enum.reduce_while` and half if normalise_dtype returns nil, doing this traversal only once.

Reply:
Done here: [53e3e87](https://github.com/elixir-explorer/explorer/pull/756/commits/53e3e8727369475c78c9242267dcc1867c0192f9)